RFS/src/test/java/com/rfs/cms/TransactionalOpenSearchDataStoreTest.java

```diff
@@ -0,0 +1,239 @@
+ * in all non-contentious cases, clients only need to make one call per work item.  Multiple calls are only
+ * required when a lease has expired and a new one is being granted since the worker/client needs to make the
+ * GET call to find out the new expiration value.
+ */
+@Slf4j
+public class TransactionalOpenSearchDataStoreTest {
+    private static final String INDEX_NAME = ".migrations_working_state";
+
+    final static OpensearchContainer<?> container =
+            new OpensearchContainer<>("opensearchproject/opensearch:2.11.0");
+
+    @BeforeAll
+    static void setupOpenSearchContainer() throws Exception {
+        // Start the container. This step might take some time...
+        container.start();
+
+        var body = "{\n" +
+                "  \"settings\": {\n" +
+                "   \"index\": {" +
+                "    \"number_of_shards\": 1,\n" +
+                "    \"number_of_replicas\": 1\n" +
+                "   }\n" +
+                "  }\n" +
+                "}\n";
+        final var httpCreateIndex = new HttpPut(container.getHttpHostAddress() + "/" + INDEX_NAME);
+        httpCreateIndex.setHeader("Content-Type", "application/json");
+        httpCreateIndex.setHeader("Accept-Encoding", "identity");
+        httpCreateIndex.setEntity(new StringEntity(body));
+
+        try (CloseableHttpClient client = HttpClients.createDefault()) {
+            var responseBody1 = client.execute(httpCreateIndex, r -> {
+                Assertions.assertEquals(HttpStatus.SC_OK, r.getCode());
+                return r.getEntity();
+            });
+        }
+    }
+
+    static String updateLeaseBodyTemplate = "{\n" +
+            "  \"scripted_upsert\": true,\n" +
+            "  \"upsert\": {\n" +
+            "    \"scriptVersion\": \"{SCRIPT_VERSION}\",\n" +
+            "    \"expiration\": 0,\n" +
+            "    \"workerId\": \"{WORKER_ID}\",\n" +
+            "    \"numAttempts\": 0\n" +
+            "  },\n" +
+            "  \"script\": {\n" +
+            "    \"lang\": \"painless\",\n" +
+            "    \"params\": { \n" +
+            "      \"clientTimestamp\": {CLIENT_TIMESTAMP},\n" +
+            "      \"expirationWindow\": {EXPIRATION_WINDOW},\n" +
+            "      \"workerId\": \"{WORKER_ID}\"\n" +
+            "    },\n" +
+            "    \"source\": \"" +
+            "      if (ctx._source.scriptVersion != \\\"{SCRIPT_VERSION}\\\") {" +
+            "        throw new IllegalArgumentException(\\\"scriptVersion mismatch.  Not all participants are using the same script: sourceVersion=\\\" + ctx.source.scriptVersion);" +
+            "      } " +
+            "      long serverTimeSeconds = System.currentTimeMillis() / 1000;" +
+            "      if (Math.abs(params.clientTimestamp - serverTimeSeconds) > {CLOCK_DEVIATION_SECONDS_THRESHOLD}) {" +
+            "        throw new IllegalArgumentException(\\\"The current times indicated between the client and server are too different.\\\");" +
```

Comment:
Why do they need to be the same at all?  The server could be different from the client in a way that would be considered 'wrong' but as long as its _consistent_ it satisfies our needs, no?

Reply:
Yes, but... right now, we're not doing any kind of adjustments or calibrations to use the server side clock when calculating when we should be killing the process.  This is sortof a poor substitution that hopefully adds some sanity.  I would hope that some day we'll get better time coordination, but the truth is that we're probably at a much coarser granularity than what even the worst offenders of clock skew would exhibit for this to be a problem.